### PR TITLE
Short circuit diffing if old array or new array is empty

### PR DESCRIPTION
**Pull request by rjchatfield**
## Changes in this pull request

Avoid calculating moves and updates.
- If `oldArray` is empty, return insert new rows.
- If `newArray` is empty, return deletes for old rows

Note: This refactor defines local properties and blocks earlier in function. But no logic was changed in these were changed.

## Pull request checklist

- [X] All tests pass. Demo project builds and runs.
- [X] I added tests, an experiment, or detailed why my change isn't tested.
- [X] I added an entry to the `CHANGELOG.md` for any breaking changes, enhancements, or bug fixes.
- [X] I have reviewed the [contributing guide](https://github.com/Instagram/IGListKit/blob/master/.github/CONTRIBUTING.md)

### CODE REVIEW

**Comment by rnystrom** [comment]
Interesting! @rjchatfield what's the inspiration behind this? Did you find a bottleneck when diffing old/new arrays?

File: CHANGELOG.md [hunk 1]
```diff
@@ -20,6 +20,10 @@ This release closes the [3.0.0 milestone](https://github.com/Instagram/IGListKit
                         withObject:(id)object;
 ```
 
+### Enhancements
+
+- Short circuit diffing if old array or new array is empty.
```
Comment (jessesquires) [inline CHANGELOG.md]: this should go under 2.2.0

Comment (rjchatfield) [inline CHANGELOG.md]: Fixed!

**Comment by rjchatfield** [comment]
@rnystrom We had our own diffing implementation which included this optimistic short circuit. It was a pretty common case to go from an empty list to a full list (init state -> loading cached state). So it just made sense for us to improve it, and avoid looping over every item and looking it up in a lookup table.

**Comment by jessesquires** [comment]
✨ friendly ping ✨ 

what's the status of this @rnystrom ?

if we don't want to accept, we should close.

**Comment by rnystrom** [comment]
Going to close for now since IMO we should have as little path-forking as possible. If someone can prove me wrong w/ some profiling on how this improves performance please do!

**Pull request closed without merging**